addons/stock/models/stock_quant.py

```diff
@@ -1238,16 +1259,46 @@ def _check_serial_number(self, product_id, lot_id, company_id, source_location_i
 
     def _move_quants(self, location_dest_id=False, message=False):
         """ Directly move a stock.quant to another location by creating a stock.move. """
-        if not location_dest_id:
-            return
+        new_quant_list = {}
         for quant in self:
             move_vals = quant._get_inventory_move_values(quant.quantity, quant.location_id, location_dest_id)
             move_vals.update({
                 'name': message or 'Quantity Relocated',
             })
-            moves = self.env['stock.move'].with_context(inventory_mode=False).create(move_vals)
+            moves = self.env['stock.move'].create(move_vals)
             moves._action_done()
+            new_quant_list[quant.id] = self._gather(quant.product_id, location_dest_id)
+        return new_quant_list
 
+    def _get_package_dict(self):
+        package_dict = {False: []}
+        for quant in self:
+            if quant.package_id and all(q in self for q in quant.package_id.quant_ids):
+                if quant.package_id.id not in package_dict.keys():
+                    package_dict[quant.package_id.id] = [quant.id]
+                else:
+                    package_dict[quant.package_id.id] += [quant.id]
+            else:
+                package_dict[False] += [quant.id]
+            quant.write({'package_id': False})
+        return package_dict
+
+    def _unreserve_quants(self):
+        for quant in self:
+            potential_sml_to_unreserve = self.env['stock.move.line'].search([
+                ('state', 'not in', ['done', 'cancel']),
+                ('product_id', '=', quant.product_id.id),
+                ('lot_id', '=', quant.lot_id.id),
+                ('location_id', '=', quant.location_id.id),
+                ('owner_id', '=', quant.owner_id.id),
+                ('package_id', '=', quant.package_id.id),
+                ('reserved_qty', '>', 0.0),
+            ])
+            if sum(potential_sml_to_unreserve.mapped('reserved_qty')) != quant.reserved_quantity:
+                raise UserError(_('The total amount to unreserve does not match the reserved amount on the quantity moved.'))
+            moves_to_recompute = potential_sml_to_unreserve.move_id
+            potential_sml_to_unreserve.reserved_uom_qty = 0
+            moves_to_recompute._recompute_state()
```

Comment:
Should use a read_group and a filter in python to avoid a huge sql amount of queries. Recompute and write should also be batched

Reply:
done

Reply:
done